README.md
For most things a simple `setTimeout()` or `setInterval()` is enough - but Javascript timers get throttled when the page is minimised/tabbed out of and you will end up with imprecise timings in the long run. This is where AloTimer comes in - you set a timer and poll it with as much precision as you need - while individual polling calls might get throttled, the end result will always be accurate.

[![Coverage Status](https://coveralls.io/repos/github/Alorel/alo-timer/badge.svg?branch=master)](https://coveralls.io/github/Alorel/alo-timer?branch=master)
[![Build Status](https://travis-ci.org/Alorel/alo-timer.svg?branch=master)](https://travis-ci.org/Alorel/alo-timer)
[![Deps](https://david-dm.org/alorel/alo-timer.svg)](https://david-dm.org/alorel/alo-timer#info=dependencies&view=list)
[![Deps](https://david-dm.org/alorel/alo-timer/dev-status.svg)](https://david-dm.org/alorel/alo-timer#info=devDependencies&view=list)


[![NPM](https://nodei.co/npm/alo-timer.png?downloads=true&downloadRank=true&stars=true)](https://www.npmjs.com/package/alo-timer)

# Functionality
 - Add or subtract any amount of miliseconds/seconds/minutes/hours/days
 - Get the amount of any of the above left individually
 - Output time left as a DD:HH:mm:ss string (customisable)
 - A simple check whether the timeout has finished
 - Pause and unpause the timer

# Installation
```html
<script src="path-to-alotimer.min.js"></script>
```
Or, if you prefer Node,
```
npm install alo-timer --save
```

# Usage example

```javascript
var span       = document.getElementById("my-countdown"),
    timer      = new AloTimer(3600000, ["hours", "minutes", "seconds"]), // 1 hr
    intervalCb = function () {
       if (!timer.hasFinished) {
           span.innerText = timer.toString();
       } else {
           span.innerText = "YOUR SCHNITZEL IS DONE!";
           clearInterval(interval);
       }
   },
   interval   = setInterval(intervalCb, 1000);
```

----------

More information:

   - [Demo](https://alorel.github.io/alo-timer)
   - [API documentation](https://alorel.github.io/alo-timer/jsdoc)
   - [Test coverage](https://coveralls.io/github/Alorel/alo-timer?branch=master)

changelog.md
1.1
=====
Pause and unpause functionality added

test/test-timer.js
var origWarn = console.warn,
    warnings = [],
    /** 18d 7h 3m 11s 79ms */
    TIME = 79 + 11000 + (60000 * 3) + (3600000 * 7) + (3600000 * 24 * 18);

console.warn = function (msg) {
    warnings.push(msg);
    origWarn.apply(console, [msg]);
};

it("Defined", function () {
    expect(AloTimer).toBeDefined();
});

describe("Adders", function () {
    var t;

    beforeEach(function () {
        t = new AloTimer(5);
    });

    it("ms", function () {
        expect(t.addMS(10) instanceof AloTimer).toBe(true);
        expect(t.timeout).toBe(15);
        expect(t.msLeft > 10).toBe(true);
        expect(t.secondsLeft).toBe(0);
        expect(t.minutesLeft).toBe(0);
        expect(t.hoursLeft).toBe(0);
        expect(t.daysLeft).toBe(0);
    });

    it("s", function () {
        expect(t.addSeconds(15) instanceof AloTimer).toBe(true);
        expect(t.timeout).toBe(15005);
        expect(t.msLeft > 0).toBe(true);
        expect(t.secondsLeft >= 14).toBe(true);
        expect(t.minutesLeft).toBe(0);
        expect(t.hoursLeft).toBe(0);
        expect(t.daysLeft).toBe(0);
    });

    it("m", function () {
        expect(t.addMinutes(15) instanceof AloTimer).toBe(true);
        expect(t.timeout).toBe(5 + (60000 * 15));
        expect(t.msLeft > 0).toBe(true);
        expect(t.secondsLeft === 0 || t.secondsLeft >= 58).toBe(true);
        expect(t.minutesLeft >= 14).toBe(true);
        expect(t.hoursLeft).toBe(0);
        expect(t.daysLeft).toBe(0);
    });

    it("h", function () {
        expect(t.addHours(15) instanceof AloTimer).toBe(true);
        expect(t.timeout).toBe(5 + (3600000 * 15));
        expect(t.msLeft > 0).toBe(true);
        expect(t.secondsLeft === 0 || t.secondsLeft >= 58).toBe(true);
        expect(t.minutesLeft === 0 || t.minutesLeft === 59).toBe(true);
        expect(t.hoursLeft >= 14).toBe(true);
        expect(t.daysLeft).toBe(0);
    });

    it("d", function () {
        expect(t.addDays(15) instanceof AloTimer).toBe(true);
        expect(t.timeout).toBe(5 + (3600000 * 24 * 15));
        expect(t.msLeft > 0).toBe(true);
        expect(t.secondsLeft === 0 || t.secondsLeft >= 58).toBe(true);
        expect(t.minutesLeft === 0 || t.minutesLeft === 59).toBe(true);
        expect(t.hoursLeft === 0 || t.hoursLeft === 23).toBe(true);
        expect(t.daysLeft >= 14).toBe(true);
    });
});

describe("Subbers", function () {
    var t;

    beforeEach(function () {
        t = new AloTimer(TIME);
    });

    it("ms", function () {
        expect(t.msLeft >= 70).toBe(true);
        expect(t.subMS(20) instanceof AloTimer).toBe(true);
        expect(t.msLeft <= 59).toBe(true);
    });

    it("s", function () {
        expect(t.secondsLeft).toBe(11);
        expect(t.subSeconds(1) instanceof AloTimer).toBe(true);
        expect(t.secondsLeft).toBe(10);
    });

    it("m", function () {
        expect(t.minutesLeft).toBe(3);
        expect(t.subMinutes(1) instanceof AloTimer).toBe(true);
        expect(t.minutesLeft).toBe(2);
    });

    it("h", function () {
        expect(t.hoursLeft).toBe(7);
        expect(t.subHours(1) instanceof AloTimer).toBe(true);
        expect(t.hoursLeft).toBe(6);
    });

    it("d", function () {
        expect(t.daysLeft).toBe(18);
        expect(t.subDays(1) instanceof AloTimer).toBe(true);
        expect(t.daysLeft).toBe(17);
    });
});

it("hasFinished & diff", function (done) {
    var t = new AloTimer(5);
    expect(t.hasFinished).toBe(false);
    setTimeout(function () {
        try {
            expect(t.hasFinished).toBe(true);
            expect(t.diff >= 6).toBe(true);
            ["ms", "seconds", "minutes", "hours", "days"].forEach(function (v) {
                expect(t[v + "Left"]).toBe(0);
            });
        } finally {
            done();
        }
    }, 6);
});

it(".timeout=", function () {
    var t = new AloTimer(500);
    t.timeout = 1000;
    expect(t.timeout).toBe(1000);

    expect(function () {
        t.timeout = "foo";
    }).toThrow(new Error("The timeout must be numeric!"));

    t = new AloTimer();
    expect(t.timeout).toBe(0);
});

describe("toString", function () {
    it("default", function () {
        expect((new AloTimer(TIME)).toString()).toBe("18:07:03:11");
    });

    it("custom", function () {
        expect((new AloTimer(TIME, ["minutes", "seconds"])).toString()).toBe("03:11");

        var t = parseInt((new AloTimer(TIME, ["ms"])).toString());
        expect(t > 70 && t <= 79).toBe(true);
    });
});

it("pause", function (done) {
    var t = new AloTimer(500);
    expect(t.pauseTime).toBe(false);
    expect(t.isPaused).toBe(false);
    expect(t.pause() instanceof AloTimer).toBe(true);
    expect(t.isPaused).toBe(true);
    expect(t.pauseTime.toString()).toBe(new Date().toString());

    t.pause();
    var idx = warnings.indexOf("The timer is already paused");

    expect(idx).not.toBe(-1);
    delete warnings[idx];

    setTimeout(function () {
        try {
            expect(t.msSincePause >= 100).toBe(true);
            expect(t.unpause() instanceof AloTimer).toBe(true);
            expect(t.msSincePause).toBe(0);
            expect(t.timeout >= 600).toBe(true);

            t.unpause();
            var idx = warnings.indexOf("The timer isn't paused");
            expect(idx).not.toBe(-1);
            delete warnings[idx];
        } finally {
            done();
        }
    }, 100);
});

src/alotimer.js
var AloTimer = (function (TIME_CHAIN, floor, Date, Error) {
    // Note that this is not global if require()'d in NodeJS, only plain Javascript
    /**
     * Creates the timer
     * @author Art <a.molcanovas@gmail.com>
     * @param {number} [timeout=0] How many milliseconds to have the timeout for
     * @param {string[]} [timeChain=["days", "hours", "minutes", "seconds"]] Which units to include in toString().
     * Valid values are "days", "hours", "minutes", "seconds", and "ms". With the default setting the output for 1
     * hour 53 minutes 11 seconds and 6 milliseconds would be 00:01:53:11.
     * @class
     * @global
     */
    var AloTimer = function (timeout, timeChain) {
        /**
         * How long the timeout is set for
         * @type {number}
         */
        this.timeout = timeout || 0;

        /**
         * When the timeout started
         * @type {number}
         * @readonly
         */
        this.timeStart = new Date().getTime();

        /**
         * Which units to include in toString()
         * @type {string[]}
         */
        this.timeChain = timeChain || TIME_CHAIN;

        /**
         * When the timer was paused. Will hold false if the timer isn't currently paused.
         * @type {Date|boolean}
         */
        this.pauseTime = false;
    }, padLeft = function (str, char, length) {
        while (str.length < length) {
            str = char + str;
        }

        return str;
    };

    /** @global */
    AloTimer.prototype = {

        /**
         * Adds milliseconds to the timer
         * @author Art <a.molcanovas@gmail.com>
         * @param val The amount to add
         * @returns {AloTimer}
         */
        addMS: function (val) {
            this.timeout += val;
            return this;
        },


        get timeout() {
            return this._timeout;
        },

        set timeout(num) {
            if (!isNaN(num)) {
                this._timeout = parseInt(num);
            } else {
                throw new Error("The timeout must be numeric!");
            }
        },

        /**
         * Pauses the timer
         * @author Art <a.molcanovas@gmail.com>
         * @returns {AloTimer}
         */
        pause: function () {
            if (this.isPaused) {
                console.warn("The timer is already paused");
            } else {
                this.pauseTime = new Date();
            }
            return this;
        },

        /**
         * Checks if the timer is currently paused
         * @author Art <a.molcanovas@gmail.com>
         * @returns {boolean}
         */
        get isPaused() {
            return this.pauseTime instanceof Date;
        },

        /**
         * Unpauses the timer
         * @author Art <a.molcanovas@gmail.com>
         * @returns {AloTimer}
         */
        unpause: function () {
            if (!this.isPaused) {
                console.warn("The timer isn't paused");
            } else {
                this.addMS(this.msSincePause);
                this.pauseTime = false;
            }
            return this;
        },

        /**
         * Returns the number of milliseconds that have passed since the timer was paused
         * @author Art <a.molcanovas@gmail.com>
         * @returns {number}
         */
        get msSincePause() {
            if (this.isPaused) {
                return new Date().getTime() - this.pauseTime;
            } else {
                return 0;
            }
        },

        /**
         * Adds seconds to the timer
         * @author Art <a.molcanovas@gmail.com>
         * @param val The amount to add
         * @returns {AloTimer}
         */
        addSeconds: function (val) {
            this.timeout += val * 1000;
            return this;
        },

        /**
         * Adds minutes to the timer
         * @author Art <a.molcanovas@gmail.com>
         * @param val The amount to add
         * @returns {AloTimer}
         */
        addMinutes: function (val) {
            this.timeout += val * 60000;
            return this;
        },

        /**
         * Adds hours to the timer
         * @author Art <a.molcanovas@gmail.com>
         * @param val The amount to add
         * @returns {AloTimer}
         */
        addHours: function (val) {
            this.timeout += val * 3600000;
            return this;
        },

        /**
         * Adds days to the timer
         * @author Art <a.molcanovas@gmail.com>
         * @param val The amount to add
         * @returns {AloTimer}
         */
        addDays: function (val) {
            this.timeout += val * 86400000;
            return this;
        },

        /**
         * Subtracts milliseconds from the timer
         * @author Art <a.molcanovas@gmail.com>
         * @param val The amount to add
         * @returns {AloTimer}
         */
        subMS: function (val) {
            this.timeout -= val;
            return this;
        },

        /**
         * Subtracts seconds from the timer
         * @author Art <a.molcanovas@gmail.com>
         * @param val The amount to add
         * @returns {AloTimer}
         */
        subSeconds: function (val) {
            this.timeout -= val * 1000;
            return this;
        },

        /**
         * Subtracts minutes from the timer
         * @author Art <a.molcanovas@gmail.com>
         * @param val The amount to add
         * @returns {AloTimer}
         */
        subMinutes: function (val) {
            this.timeout -= val * 60000;
            return this;
        },

        /**
         * Subtracts hours from the timer
         * @author Art <a.molcanovas@gmail.com>
         * @param val The amount to add
         * @returns {AloTimer}
         */
        subHours: function (val) {
            this.timeout -= val * 3600000;
            return this;
        },

        /**
         * Subtracts days from the timer
         * @author Art <a.molcanovas@gmail.com>
         * @param val The amount to add
         * @returns {AloTimer}
         */
        subDays: function (val) {
            this.timeout -= val * 86400000;
            return this;
        },

        /**
         * Checks whether the timeout has finished
         * @author Art <a.molcanovas@gmail.com>
         * @returns {boolean}
         */
        get hasFinished() {
            return (new Date().getTime() - this.timeStart) >= this.timeout;
        },

        /**
         * Returns the amount of milliseconds remaining
         * @author Art <a.molcanovas@gmail.com>
         * @returns {number}
         */
        get msLeft() {
            var diff = this.diff;
            return diff < this.timeout ? (this.timeout - diff) % 1000 : 0;
        },

        /**
         * Returns the difference between timeStart and now, taking pause time into account
         * @author Art <a.molcanovas@gmail.com>
         * @returns {number}
         */
        get diff() {
            return new Date().getTime() - this.timeStart - this.msSincePause;
        },

        /**
         * Returns the amount of seconds remaining
         * @author Art <a.molcanovas@gmail.com>
         * @returns {number}
         */
        get secondsLeft() {
            var diff = this.diff;
            return diff < this.timeout ? floor(((this.timeout - diff) / 1000) % 60) : 0;
        },

        /**
         * Returns the amount of minutes remaining
         * @author Art <a.molcanovas@gmail.com>
         * @returns {number}
         */
        get minutesLeft() {
            var diff = this.diff;
            return diff < this.timeout ? floor(((this.timeout - diff) / 60000) % 60) : 0;
        },

        /**
         * Returns the amount of hours remaining
         * @author Art <a.molcanovas@gmail.com>
         * @returns {number}
         */
        get hoursLeft() {
            var diff = this.diff;
            return diff < this.timeout ? floor(((this.timeout - diff) / 3600000) % 24) : 0;
        },

        /**
         * Returns the amount of days remaining
         * @author Art <a.molcanovas@gmail.com>
         * @returns {number}
         */
        get daysLeft() {
            var diff = this.diff;
            return diff < this.timeout ? floor((this.timeout - diff) / 86400000) : 0;
        },

        /**
         * Returns a string representation of the amount of time remaining
         * @author Art <a.molcanovas@gmail.com>
         * @returns {string}
         */
        toString: function () {
            var arr = [], tmp, i;

            for (i = 0; i < this.timeChain.length; i++) {
                tmp = this[this.timeChain[i] + "Left"];

                arr.push(padLeft(this[this.timeChain[i] + "Left"].toString(), "0", this.timeChain[i] === "ms" ? 3 : 2));
            }

            return arr.join(":");
        }
    };

    return AloTimer;
})(['days', 'hours', 'minutes', 'seconds'], Math.floor, Date, Error);

if (typeof module !== "undefined" && typeof module.exports !== "undefined") {
    module.exports = AloTimer;
}